File: Hard/2026-01-28-3651-Minimum-Cost-Path-with-Teleportations/solution.py
import heapq

class Solution:
    def minCost(self, grid: List[List[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])
        directions = [(0, 1), (1, 0)]
        min_cost = [[float('inf')] * (k + 1) for _ in range(m)]
        min_cost[0][0] = 0
        pq = [(0, 0, 0)]  # (cost, x, y)

        while pq:
            cost, x, y = heapq.heappop(pq)

            if (x, y) == (m - 1, n - 1):
                return cost

            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                if 0 <= nx < m and 0 <= ny < n:
                    new_cost = cost + grid[nx][ny]
                    if new_cost < min_cost[nx][ny]:
                        min_cost[nx][ny] = new_cost
                        heapq.heappush(pq, (new_cost, nx, ny))

            if k > 0:
                for i in range(m):
                    for j in range(n):
                        if grid[i][j] <= grid[x][y]:
                            if cost < min_cost[i][j]:
                                min_cost[i][j] = cost
                                heapq.heappush(pq, (cost, i, j))

        return -1

File: Hard/2026-01-28-3651-Minimum-Cost-Path-with-Teleportations/solution.js
class Solution {
    minCost(grid, k) {
        const m = grid.length, n = grid[0].length;
        const INF = Infinity;
        const dp = Array.from({ length: m }, () => Array.from({ length: n }, () => Array(k + 1).fill(INF)));
        dp[0][0][k] = 0;

        const directions = [[1, 0], [0, 1]];

        for (let i = 0; i < m; i++) {
            for (let j = 0; j < n; j++) {
                for (let t = 0; t <= k; t++) {
                    if (dp[i][j][t] === INF) continue;

                    for (const [dx, dy] of directions) {
                        const ni = i + dx, nj = j + dy;
                        if (ni < m && nj < n) {
                            dp[ni][nj][t] = Math.min(dp[ni][nj][t], dp[i][j][t] + grid[ni][nj]);
                        }
                    }

                    if (t > 0) {
                        const teleportCost = grid[i][j];
                        for (let x = 0; x < m; x++) {
                            for (let y = 0; y < n; y++) {
                                if (grid[x][y] <= teleportCost) {
                                    dp[x][y][t - 1] = Math.min(dp[x][y][t - 1], dp[i][j][t]);
                                }
                            }
                        }
                    }
                }
            }
        }

        let minCost = INF;
        for (let t = 0; t <= k; t++) {
            minCost = Math.min(minCost, dp[m - 1][n - 1][t]);
        }

        return minCost;
    }
}